crates/doc/src/inference.rs

```diff
@@ -1484,6 +1486,201 @@ impl Shape {
     }
 }
 
+const MAX_ROOT_FIELDS: usize = 750;
+const MAX_NESTED_FIELDS: usize = 200;
+
+impl Shape {
+    // Widen a Shape to make the provided AsNode value fit.
+    // Returns a hint if some locations might exceed their maximum allowable size.
+    // NOTE: If a particular location defines `additionalProperties` as a subschema, don't
+    // add the field from `AsNode`, instead jump straight to squashing the the field into `additionalProperties`
+    fn widen<'n, N>(&mut self, node: &'n N) -> bool
+    where
+        N: AsNode,
+    {
+        self.widen_inner(node, Location::Root)
+    }
+
+    fn widen_inner<'n, N>(&mut self, node: &'n N, loc: Location) -> bool
+    where
+        N: AsNode,
+    {
+        use crate::{Field, Fields};
+
+        match node.as_node() {
+            crate::Node::Object(obj) => {
+                self.type_ = self.type_ | types::OBJECT;
```

Comment:
should this be an impl of ObjShape ? There's a lot happening / a lot of cyclomatic nesting in this routine

Reply:
Yup, way better. Moved 👍